nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/utils/AzureStorageUtils.java

```diff
@@ -85,6 +85,22 @@
             .sensitive(true)
             .build();
 
+    public static final PropertyDescriptor ENDPOINT_SUFFIX = new PropertyDescriptor.Builder()
+            .name("storage-endpoint-suffix")
+            .displayName("Storage Endpoint Suffix")
+            .description(
+                    "Storage accounts in public Azure always use a common FQDN suffix. " +
+                    "Override this endpoint suffix with a different suffix in certain circumsances (like Azure Stack or non-public Azure regions). " +
+                    "The preferred way is to configure them through a controller service specified in the Storage Credentials property. " +
+                    "The controller service can provide a common/shared configuration for multiple/all Azure processors. Furthermore, the credentials " +
+                    "can also be looked up dynamically with the 'Lookup' version of the service.")
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
```

Comment:
I think this is a good observation. So the question turns into whether to assume a full URL, just the suffix, or both (easy enough to detect the difference).

Reply:
This PR addresses the need to support Azure Stack and Azure Gov cloud.

This PR binds to 
CloudStorageAccount.CloudStorageAccount(
            final StorageCredentials storageCredentials, final boolean useHttps, final String endpointSuffix, String accountName)  

API, which has equivalent common pattern and endpoint suffix support in both Datalake SDK and Storage SDK.

While, to bind the development environment, I see different SDK API.

CloudStorageAccount.getDevelopmentStorageAccount(final URI proxyUri)

, which uses the fixed DEVSTORE_ACCOUNT_NAME and DEVSTORE_ACCOUNT_KEY.

    private static final String DEVSTORE_ACCOUNT_KEY = "Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==";
    private static final String DEVSTORE_ACCOUNT_NAME = "devstoreaccount1";

This means that you don't even need to set account name and key to bind to the development environment.

@esecules and @jfrazee , 
How critical is it to address NIFI-7386 in this PR? We could address the problem with 'development-environment-uri', rather than endpoint-suffix. 

And, using one property for endpoint_suffix in one case and for endpoint URI in another case will make it confusing, in my opinion.

Reply:
Given https://docs.microsoft.com/en-us/azure/storage/common/storage-configure-connection-string#specify-an-http-proxy, the property for NIFI-7386 would be 'DevelopmentStorageProxyUri' or 'development-storage-proxy-uri' in my opinion and we should call the correct API in this case(i.e. CloudStorageAccount.getDevelopmentStorageAccount(final URI proxyUri)).

And, "EndPointSuffix" is one of connection property showing in Azure portal, and thus this is the obvious configuration in non public Azure cloud environemnt.

Reply:
@sjyang18 I would love to have NIFI-7386 addressed here but I will defer to someone affiliated with Apache ( @jfrazee ) as I am the user who logged the bug in the first place.

I wonder, is there a finite list of possible endpoints? If so the endpoint suffix could maybe be a dropdown list. There'd still need to be a text field for the storage emulator though so maybe it'll not be intuitive.

Reply:
@esecules, @jfrazee , let me work on a patch for NIFI-7386 and let's review.

Reply:
I opened a new PR for NIFI-7386. https://github.com/apache/nifi/pull/4286.

Reply:
I think the precedent for these is to use a dropdown (`allowableValues()`) and there is a list of these in [`client-runtime`](https://github.com/Azure/autorest-clientruntime-for-java/blob/23e3142db1f7fbcd8a871ed791c1a806285ee81c/azure-client-runtime/src/main/java/com/microsoft/azure/AzureEnvironment.java#L144), which we already pull in, but it doesn't include Azure Stack so we'd need to add that. Are there any (un)expected drawbacks to doing this?

Reply:
After thinking about it more, there is a use case that can't be satisfied using `allowableValues()`. The storage account, container, etc. can be pulled from FlowFile attributes so if the suffix is freeform, you can have the same processor(s) connecting across commercial and gov clouds based on attributes.

I lean toward `allowableValues()` nonetheless because I'd be a little surprised if it's a common use case and it can be satisfied by just using two processors.

Interested to hear from others on this though.